cartocrow/renderer/svg_renderer.h

```diff
@@ -0,0 +1,118 @@
+#define CARTOCROW_RENDERER_SVG_RENDERER_H
+
+#include <QPainter>
+
+#include <filesystem>
+#include <fstream>
+#include <stack>
+
+#include "geometry_painting.h"
+#include "geometry_renderer.h"
+
+namespace cartocrow::renderer {
+
+/// The style for a GeometryWidget.
+struct SvgRendererStyle {
+	/// The draw mode.
+	int m_mode = GeometryRenderer::stroke;
+	/// The diameter of points.
+	double m_pointSize = 10;
+	/// The color of points and lines.
+	std::string m_strokeColor = "#000000";
+	/// The opacity of points and lines.
+	double m_strokeOpacity = 1;
+	/// The width of lines.
+	double m_strokeWidth = 1;
+	/// Whether the width is interpreted as absolute, that is, independent of
+	/// the renderer's zoom factor.
+	bool m_absoluteWidth = false;
+	/// The color of filled shapes.
+	std::string m_fillColor = "#0066cb";
+	/// The opacity of filled shapes.
+	double m_fillOpacity = 1;
+};
+
+/// SVG specialization of the GeometryRenderer.
+class SvgRenderer : public GeometryRenderer {
+
+  public:
+	SvgRenderer() = default;
+
+	/// Constructs a SvgRenderer for the given painting.
+	SvgRenderer(const std::shared_ptr<GeometryPainting>& painting);
+	SvgRenderer(const std::shared_ptr<GeometryPainting>& painting, const std::string& name);
+
+	/// Saves the painting to an SVG file with the given name.
+	void save(const std::filesystem::path& file);
+
+	void draw(const Point<Inexact>& p) override;
+	void draw(const PolygonWithHoles<Inexact>& p) override;
+	void draw(const Circle<Inexact>& c) override;
+	void draw(const BezierSpline& s) override;
+	void draw(const Line<Inexact>& l) override;
+	void draw(const Ray<Inexact>& r) override;
+	void draw(const RenderPath& p) override;
+	void drawText(const Point<Inexact>& p, const std::string& text) override;
+
+	void pushStyle() override;
+	void popStyle() override;
+	void setMode(int mode) override;
+	void setStroke(Color color, double width, bool absoluteWidth = false) override;
+	void setStrokeOpacity(int alpha) override;
+	void setFill(Color color) override;
+	void setFillOpacity(int alpha) override;
+
+	void addPainting(const std::shared_ptr<GeometryPainting>& painting);
+	void addPainting(const std::shared_ptr<GeometryPainting>& painting, const std::string& name);
+
+  private:
+	/// The output file we're writing to.
+	std::ofstream m_out;
+	/// Converts a polygon to an SVG path specification.
+	std::string convertPolygonToCurve(const Polygon<Inexact>& p) const;
+	/// Returns the style as a string that can be added to an SVG path element.
+	std::string getStyle() const;
+	/// Returns the style as a string that can be added to an SVG use element.
+	std::string getVertexStyle() const;
+	/// Escapes SVG's reserved characters.
+	std::string escapeForSvg(const std::string& text) const;
+
+	struct DrawnPainting {
+		/// The painting itself.
+		std::shared_ptr<GeometryPainting> m_painting;
+		/// The name of the painting displayed as a layer name in ipe.
+		std::optional<std::string> name;
```

Comment:
The comment still talks about ipe. I'm wondering whether we should even pass a name? Since paintings becomes groups. Is there some naming format for SVG elements so that it shows up e.g. Inkscape?

Reply:
Right, good point. I was going to look into this, and then I didn't. 

SVG doesn't have the concept of layers (that's why I just made groups). One can define the name of groups, but I don't think that'll show up in any editor. 

Inkscape does have layers somehow. I guess they have some extended format that supports it. I'll do some experimentats with Inkscape files to try and reverse engineer how it works.

Reply:
Fixed, it turned out to be quite simple (`<g inkscape:groupmode="layer" inkscape:label="whatever name you want">`).